ENIGMAsystem/SHELL/Platforms/Win32/WINDOWScallback.cpp

```diff
@@ -200,9 +200,11 @@ namespace enigma
         keybdstatus[key]=1;
         if (key!=18)
         {
-          if ((lParam&(1<<29))>0)
-               last_keybdstatus[18]=keybdstatus[18], keybdstatus[18]=1;
-          else last_keybdstatus[18]=keybdstatus[18], keybdstatus[18]=0;
+          if ((lParam&(1<<29))>0) {
+            if (wParam == VK_F4) PostMessage(hWnd, WM_SYSCOMMAND, SC_CLOSE, 0);
```

Comment:
@JoshDreamland I could have used SendMessage instead of PostMessage but that API blocks until WM_CLOSE returns, if it even returns at all. That sounds like a bad idea right? And is it ok if I use the UPPERCASE virtual key constants here since they don't conflict with the user ones?

Reply:
Using the uppercase constants is correct. We should be able to change our constants without breaking the Win32 system. It's happenstance that our constants have the same value as the GM constants which use the Win32 constants.

Reply:
Ok, I can't use a constant for the 29 that fundies mentioned though because that's just the 29th bit of lParam in accordance with the WM_SYSKEYDOWN message, it doesn't correspond to the `VK_ALT` constant. However, I can replace 18 with the correct constant.
https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-syskeydown

Reply:
I don't see reason you can't use a const for 29. Why are you getting the 29th bit? A variable name and comment should reflect that

Reply:
Open the link in my last comment fundies, look at `lParam` and see what it says for the 29th bit. Anyway, regardless, I changed this back due to misunderstanding, I can fix the other thing in a separate pr as I have related input changes to make yet.

Reply:
Well from breifly scanning docs 29 looks like it might be equiv to KF_ALTDOWN which is the manipulator so if you cant get the value from that then just const int KF_ALTDOWN_BIT = 29; or whatever so it's clear what 29 is

Reply:
You are correct, nice find. I'll commit it just to make you happy.

HIWORD returns the high order 16 bit word of the 32 bit DWORD lParam. The value of KF_ALTDOWN is 8192 which is 2^13 and 13 + 16 = 29th bit.
`bAltIsDown = (HIWORD(lParam) & KF_ALTDOWN);`

Reply:
There a08049d21a9ac202f1bb830ae8270c948dff539e

Reply:
The code was actually bugged before it seems too. The SYSKEYUP version was interpreting it backwards, but the documentation is very clear that it's only ever set when ALT is down, with the key referring to the main key being pressed, like F4.